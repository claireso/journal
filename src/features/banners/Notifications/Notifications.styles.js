import styled from 'styled-components'

export const ButtonSubscribe = styled.a.attrs(() => ({
  href: '#',
  role: 'button'
}))`
  color: inherit;
  text-decoration: none;
  &:before {
    content: '';
    display: block;
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
  }
  &:hover {
    border-bottom: 1px solid currentColor;
  }
`
