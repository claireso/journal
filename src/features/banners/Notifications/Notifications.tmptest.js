//@TODO restore tests

// import { render, fireEvent, waitFor } from '@testing-library/react'

// import { TranslationsProvider } from '@services/translations/hooks/useTranslations'
// import { getTranslations } from '@services/translations/'
// import * as notifications from '@services/notifications'

// import Notifications from './index'

// describe('<Notifications />', () => {
//   const renderComponent = () =>
//     render(
//       <TranslationsProvider translations={getTranslations('en', 'client')}>
//         <Notifications />
//       </TranslationsProvider>
//     )

//   beforeEach(() => {
//     global.setNotificationPermission()
//   })

//   test('should render component', async () => {
//     const spy = jest.spyOn(notifications, 'getSubscription').mockImplementation(() => Promise.resolve(null))

//     const { container } = renderComponent()

//     await waitFor(() => expect(spy).toHaveBeenCalled())

//     expect(container).toMatchSnapshot()
//     spy.mockRestore()
//   })

//   test('should not render component (user has already subscribed)', async () => {
//     const spy = jest.spyOn(notifications, 'getSubscription').mockImplementation(() => Promise.resolve({}))

//     const { container } = renderComponent()

//     await waitFor(() => expect(spy).toHaveBeenCalled())

//     expect(container).toMatchSnapshot()
//     spy.mockRestore()
//   })

//   test('should not render component (user has blocked notifications)', () => {
//     global.setNotificationPermission('denied')

//     const { container } = renderComponent()

//     expect(container).toMatchSnapshot()
//     global.setNotificationPermission()
//   })

//   test('should subscribe and hide banner', async () => {
//     // mock func subscribe
//     const spy = jest.spyOn(notifications, 'subscribe').mockImplementation(() => Promise.resolve())

//     // mock func getSubscription
//     const spySubscription = jest.spyOn(notifications, 'getSubscription').mockImplementation(() => Promise.resolve(null))

//     const { container, getByText } = renderComponent()

//     await waitFor(() => getByText(/^Enable notifications/i), {
//       container
//     })

//     fireEvent.click(getByText(/^Enable notifications/i))

//     await waitFor(() => expect(spy).toHaveBeenCalled())

//     expect(container).toMatchSnapshot()
//     spy.mockRestore()
//     spySubscription.mockRestore()
//   })

//   test('should not subscribe and hide banner (denied story)', async () => {
//     // mock func subscribe
//     const spy = jest.spyOn(notifications, 'subscribe').mockImplementation(() => Promise.reject())

//     // mock func getSubscription
//     const spySubscription = jest.spyOn(notifications, 'getSubscription').mockImplementation(() => Promise.resolve(null))

//     const { container, getByText } = renderComponent()

//     await waitFor(() => getByText(/^Enable notifications/i), {
//       container
//     })

//     global.setNotificationPermission('denied')

//     fireEvent.click(getByText(/^Enable notifications/i))

//     await waitFor(() => expect(spy).toHaveBeenCalled())

//     expect(container).toMatchSnapshot()
//     spy.mockRestore()
//     spySubscription.mockRestore()
//   })
// })
