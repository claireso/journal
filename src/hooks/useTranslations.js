import React from 'react'
import PropTypes from 'prop-types'

import translations from '../translations.json'

const TranslationsContext = React.createContext({})

export const TranslationsProvider = ({ children, translations }) => (
  <TranslationsContext.Provider value={translations}>{children}</TranslationsContext.Provider>
)

TranslationsProvider.propTypes = {
  children: PropTypes.node,
  translations: PropTypes.object
}

export const useTranslations = () => React.useContext(TranslationsContext)

export const getTranslations = (lang, namespace) => {
  const dict = translations[lang]

  if (!dict) {
    throw new Error(`Language '${lang}' is not supported by the application`)
  }

  if (!namespace) return dict

  const trans = dict[namespace]

  if (!trans) {
    throw new Error(`Namespace '${namespace}' is not defined for the language ${lang}`)
  }

  return trans
}
