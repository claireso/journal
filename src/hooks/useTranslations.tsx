'use client'

import React from 'react'
import PropTypes from 'prop-types'

import { getTranslations } from '../translations'

const TranslationsContext = React.createContext({})

type Translations = {
  [key: string]: object | string
}

interface TranslationsProviderProps {
  children: React.ReactNode
  namespace: 'client' | 'admin'
}

export const TranslationsProvider = ({ children, namespace }: TranslationsProviderProps) => {
  const translations = getTranslations(process.env.NEXT_PUBLIC_WEBSITE_LANGUAGE, namespace)

  return <TranslationsContext.Provider value={translations}>{children}</TranslationsContext.Provider>
}

TranslationsProvider.propTypes = {
  children: PropTypes.node,
  translations: PropTypes.object
}

export const useTranslations = (): Translations => React.useContext(TranslationsContext)
