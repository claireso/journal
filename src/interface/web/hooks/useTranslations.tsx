'use client'

import React from 'react'
import PropTypes from 'prop-types'

import { getTranslations } from '@infrastructure/translations'

const TranslationsContext = React.createContext({})

type Translations = {
  [key: string]: object | string
}

interface TranslationsProviderProps {
  children: React.ReactNode
  namespace: 'client' | 'admin'
  lang?: string
}

export const TranslationsProvider = ({ children, namespace, lang }: TranslationsProviderProps) => {
  const translations = getTranslations(lang ?? 'en', namespace)

  return <TranslationsContext.Provider value={translations}>{children}</TranslationsContext.Provider>
}

TranslationsProvider.propTypes = {
  children: PropTypes.node,
  translations: PropTypes.object
}

export const useTranslations = (): Translations => React.useContext(TranslationsContext)
