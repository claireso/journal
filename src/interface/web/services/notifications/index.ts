import logger from '@infrastructure/logger'
import urlBase64ToUint8Array from '@utils/urlBase64ToUint8Array'
import arrayBufferToUrlBase64 from '@utils/arrayBufferToUrlBase64'

const NOTIFICATIONS_PUBLIC_KEY = process.env.NEXT_PUBLIC_NOTIFICATIONS_PUBLIC_KEY

export const areDenied = () => {
  return Notification.permission === 'denied'
}

export const areGranted = () => {
  return Notification.permission === 'granted'
}

export const areDefault = () => {
  return Notification.permission === 'default'
}

export const subscribe = async (registration?: ServiceWorkerRegistration) => {
  try {
    if (!NOTIFICATIONS_PUBLIC_KEY) return

    const applicationServerKey = urlBase64ToUint8Array(NOTIFICATIONS_PUBLIC_KEY, window)

    if (!registration) {
      registration = await getRegistration()
    }

    const subscription = await registration.pushManager.subscribe({
      userVisibleOnly: true,
      applicationServerKey: applicationServerKey
    })

    await fetch('/api/subscriptions', {
      method: 'post',
      headers: {
        'Content-type': 'application/json'
      },
      body: JSON.stringify({
        subscription: subscription
      })
    })
  } catch (err) {
    logger.error(err)
  }
}

export const getRegistration = async () => {
  try {
    return await navigator.serviceWorker.ready
  } catch {
    throw new Error('Notifications: can not get registration')
  }
}

export const getSubscription = async () => {
  try {
    const registration = await getRegistration()
    if (registration) {
      const subscription = await registration.pushManager.getSubscription()
      return subscription
    }
  } catch {
    throw new Error('Notifications: can not get subscription')
  }
}

// a subscription is valid if its applicationServerKey is the same of the application notifications public key defined in environment variable
export const isSubscriptionValid = (subscription: PushSubscription): boolean => {
  const { applicationServerKey } = subscription.options
  if (!applicationServerKey) {
    return false
  }

  const applicationServerKeyString = arrayBufferToUrlBase64(applicationServerKey)

  return applicationServerKeyString === NOTIFICATIONS_PUBLIC_KEY
}
