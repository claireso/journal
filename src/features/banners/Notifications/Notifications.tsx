'use client'

import React, { useState, useEffect, useCallback } from 'react'

import { useTranslations } from '@hooks/useTranslations'
import * as notifications from '@services/notifications'

import * as S from './Notifications.styles'

import Flash from '@components/Flash'

const BannerNotifications = () => {
  const [isVisible, setIsVisible] = useState(false)
  const translations = useTranslations()

  const hideBanner = useCallback(() => setIsVisible(false), [])
  const showBanner = useCallback(() => setIsVisible(true), [])

  const subscribe = useCallback(
    async (event: React.MouseEvent<HTMLAnchorElement>) => {
      event && event.preventDefault()

      try {
        await notifications.subscribe()
        hideBanner()
      } catch (err) {
        // user decline
        if (notifications.areDenied()) {
          hideBanner()
          return
        }

        throw new Error('Banner: can not subscribe')
      }
    },
    [hideBanner]
  )

  useEffect(() => {
    // do not display banner if service worker or Notification is not support
    // do not display banner if push is not enabled
    // do not display banner in safari
    if (
      process.env.IS_NOTIFICATIONS_ENABLED === 'false' ||
      !('serviceWorker' in navigator) ||
      !('Notification' in window) ||
      'safari' in window // @TODO
    ) {
      return
    }

    // do not display banner if user has denied notifications
    if (notifications.areDenied()) {
      return
    }

    // check if user has already subscribed
    notifications
      .getSubscription()
      .then((subscription) => {
        if (!subscription) {
          showBanner()
        }
      })
      .catch(() => {
        throw new Error('Banner: can not get subscription')
      })
  }, [showBanner])

  if (!isVisible) return null

  return (
    <Flash data-testid="flash-notifications" status="info" onClose={hideBanner} css={{ mb: 0 }}>
      <S.ButtonSubscribe role="button" href="#" onClick={subscribe}>
        {/* @ts-ignore */}
        {translations.bannerNotifications}
      </S.ButtonSubscribe>
    </Flash>
  )
}

export default BannerNotifications
