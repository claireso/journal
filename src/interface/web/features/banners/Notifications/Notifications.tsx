'use client'

import React, { useState, useEffect, useCallback } from 'react'

import { useTranslations } from '@web/hooks/useTranslations'
import * as notifications from '@web/services/notifications'

import * as cls from './styles.css'

import Link from '@web/components/Links'
import Flash from '@web/components/Flash'

const BannerNotifications = () => {
  const [isVisible, setIsVisible] = useState(false)
  const translations = useTranslations()

  const hideBanner = useCallback(() => setIsVisible(false), [])
  const showBanner = useCallback(() => setIsVisible(true), [])

  const subscribe = useCallback(
    async (event: React.MouseEvent<HTMLAnchorElement>) => {
      event && event.preventDefault()

      try {
        await notifications.subscribe()
        hideBanner()
      } catch (err) {
        // user decline
        if (notifications.areDenied()) {
          hideBanner()
          return
        }

        throw new Error('Banner: can not subscribe')
      }
    },
    [hideBanner]
  )

  useEffect(() => {
    // do not display banner if service worker or Notification is not support
    // do not display banner if push is not enabled
    // do not display banner in safari
    if (
      process.env.NEXT_PUBLIC_IS_NOTIFICATIONS_ENABLED === 'false' ||
      !('serviceWorker' in navigator) ||
      !('Notification' in window) ||
      'safari' in window // @TODO
    ) {
      return
    }

    // do not display banner if user has denied notifications
    if (notifications.areDenied()) {
      return
    }

    // check if user has already subscribed
    notifications
      .getSubscription()
      .then((subscription) => {
        if (!subscription) {
          showBanner()
        }
      })
      .catch(() => {
        throw new Error('Banner: can not get subscription')
      })
  }, [showBanner])

  if (!isVisible) return null

  return (
    <Flash status="info" onClose={hideBanner}>
      <Link className={cls.link} role="button" href="#" onClick={subscribe}>
        {translations.bannerNotifications as string}
      </Link>
    </Flash>
  )
}

export default BannerNotifications
