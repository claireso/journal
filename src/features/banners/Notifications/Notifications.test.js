import { render, screen, fireEvent, waitFor } from '@testing-library/react'

import { TranslationsProvider } from '@hooks/useTranslations'
import * as notifications from '@services/notifications'

import Notifications from './index'

describe('<Notifications />', () => {
  const renderComponent = () =>
    render(
      <TranslationsProvider namespace="client">
        <Notifications />
      </TranslationsProvider>
    )

  beforeEach(() => {
    global.setNotificationPermission()
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  it('should render component', async () => {
    const getSubscription = jest.spyOn(notifications, 'getSubscription').mockImplementation(() => Promise.resolve(null))

    renderComponent()

    await waitFor(() => expect(getSubscription).toHaveBeenCalled())

    await waitFor(() => {
      expect(screen.getByText('Enable notifications to be alerted of new publication')).toBeInTheDocument()
    })
  })

  it('should not render component (user has already subscribed)', async () => {
    const getSubscription = jest.spyOn(notifications, 'getSubscription').mockImplementation(() => Promise.resolve({}))

    renderComponent()

    await waitFor(() => expect(getSubscription).toHaveBeenCalled())

    await waitFor(() => {
      expect(screen.queryByText('Enable notifications to be alerted of new publication')).not.toBeInTheDocument()
    })
  })

  it('should not render component (user has blocked notifications)', () => {
    global.setNotificationPermission('denied')

    const { asFragment } = renderComponent()

    expect(asFragment()).toMatchSnapshot()
    global.setNotificationPermission()
  })

  it('should subscribe and hide banner', async () => {
    // mock func subscribe
    const subscribe = jest.spyOn(notifications, 'subscribe').mockImplementation(() => Promise.resolve())

    // mock func getSubscription
    jest.spyOn(notifications, 'getSubscription').mockImplementation(() => Promise.resolve(null))

    renderComponent()

    await waitFor(() => screen.getByText(/^Enable notifications/i))

    fireEvent.click(screen.getByText(/^Enable notifications/i))

    await waitFor(() => expect(subscribe).toHaveBeenCalled())
    await waitFor(() => {
      expect(screen.queryByText('Enable notifications to be alerted of new publication')).not.toBeInTheDocument()
    })
  })

  it('should not subscribe and hide banner (denied story)', async () => {
    // mock func subscribe
    const subscribe = jest.spyOn(notifications, 'subscribe').mockImplementation(() => Promise.reject())

    // mock func getSubscription
    jest.spyOn(notifications, 'getSubscription').mockImplementation(() => Promise.resolve(null))

    renderComponent()

    await waitFor(() => screen.getByText(/^Enable notifications/i))

    global.setNotificationPermission('denied')

    fireEvent.click(screen.getByText(/^Enable notifications/i))

    await waitFor(() => expect(subscribe).toHaveBeenCalled())

    await waitFor(() => {
      expect(screen.queryByText('Enable notifications to be alerted of new publication')).not.toBeInTheDocument()
    })
  })
})
