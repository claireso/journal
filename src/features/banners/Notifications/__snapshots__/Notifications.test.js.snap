// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Notifications /> should not render component (user has already subscribed) 1`] = `<DocumentFragment />`;

exports[`<Notifications /> should not render component (user has blocked notifications) 1`] = `<DocumentFragment />`;

exports[`<Notifications /> should not subscribe and hide banner (denied story) 1`] = `<DocumentFragment />`;

exports[`<Notifications /> should render component 1`] = `
<DocumentFragment>
  .c1 {
  color: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c1:before {
  content: '';
  display: block;
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
}

.c1:hover {
  border-bottom: 1px solid currentColor;
}

.c3 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: none;
  border: none;
  color: currentColor;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-family: var(--font-family);
  outline: none;
  padding: 0;
  margin: 0;
}

.c4 {
  border-radius: 0.2rem;
  font-size: var(--font-size-smaller);
  line-height: 1.3333333;
  padding: 1.6rem;
  -webkit-transition: background 150ms ease-out;
  transition: background 150ms ease-out;
}

.c4 > svg {
  margin: 0 -0.4rem 0 0.4rem;
}

.c2 + .c2 {
  margin: 0 0 0 0.8rem;
}

.c0 {
  background: var(--flash-default);
  color: var(--flash-default-text);
  font-size: var(--font-size-normal);
  margin: 0 0 var(--gutter);
  padding: 2rem;
  position: relative;
  text-align: center;
  opacity: 0;
}

.c5 {
  padding: 0 2rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: absolute;
  right: 0;
  top: 0;
  bottom: 0;
}

.c5 > svg {
  fill: currentColor;
  margin: 0;
}

<div
    class="c0"
    data-testid="flash-notifications"
  >
    <a
      class="c1"
      href="#"
      role="button"
    >
      Enable notifications to be alerted of new publication
    </a>
    <button
      class="c2 c3 c4 c5"
    >
      <svg
        height="16"
        viewBox="0 0 22 28"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M20.281 20.656c0 0.391-0.156 0.781-0.438 1.062l-2.125 2.125c-0.281 0.281-0.672 0.438-1.062 0.438s-0.781-0.156-1.062-0.438l-4.594-4.594-4.594 4.594c-0.281 0.281-0.672 0.438-1.062 0.438s-0.781-0.156-1.062-0.438l-2.125-2.125c-0.281-0.281-0.438-0.672-0.438-1.062s0.156-0.781 0.438-1.062l4.594-4.594-4.594-4.594c-0.281-0.281-0.438-0.672-0.438-1.062s0.156-0.781 0.438-1.062l2.125-2.125c0.281-0.281 0.672-0.438 1.062-0.438s0.781 0.156 1.062 0.438l4.594 4.594 4.594-4.594c0.281-0.281 0.672-0.438 1.062-0.438s0.781 0.156 1.062 0.438l2.125 2.125c0.281 0.281 0.438 0.672 0.438 1.062s-0.156 0.781-0.438 1.062l-4.594 4.594 4.594 4.594c0.281 0.281 0.438 0.672 0.438 1.062z"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;

exports[`<Notifications /> should subscribe and hide banner 1`] = `<DocumentFragment />`;
